File: Algorithm/Python/1000+/1585. Check If String Is Transformable With Substring Sort Operations.py

# greedy, O(m+n) time

from collections import defaultdict
class Solution(object):
    def isTransformable(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        idx = defaultdict(list)
        for i in range(len(s)):
            idx[s[i]].append(i)
            
        counter = defaultdict(int)
        for c in t:
            if counter[c] >= len(idx[c]):
                return False
            num = ord(c) - 48
            for i in range(num):
                sc = chr(48+i)
                if counter[sc] < len(idx[sc]) and idx[sc][counter[sc]] < idx[c][counter[c]]:
                    return False
            counter[c] += 1
        return True
        
        
        

File: Algorithm/C++/1000+/1585. Check If String Is Transformable With Substring Sort Operations.cc
#include <string>
#include <vector>

// greedy, O(m+n) time

class Solution {
public:
    bool isTransformable(std::string s, std::string t) {
        std::vector<std::vector<int>> idx(10);
        std::vector<int> counter(10, 0);
        for(int i=0; i<s.size(); i++)
            idx[s[i]-'0'].push_back(i);
        for(char c : t) {
            int val = c - '0';
            if (counter[val] >= idx[val].size()) return false;
            for (int i=0; i<val; i++)
                if (counter[i] < idx[i].size() && idx[i][counter[i]] < idx[val][counter[val]])
                    return false;
            counter[val]++;
        }
        return true;
    }
};